compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2132,24 +2303,42 @@ pub struct SourceScopeLocalData {
 ///////////////////////////////////////////////////////////////////////////
 // Operands
 
-/// These are values that can appear inside an rvalue. They are intentionally
-/// limited to prevent rvalues from being nested in one another.
+/// An operand in MIR represents a "value" in Rust, the definition of which is undecided and part of
+/// the memory model. One proposal for a definition of values can be found [on UCG][value-def].
+///
+/// [value-def]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/value-domain.md
+///
+/// The most common way to create values is via a place to value conversion. A place to value
+/// conversion is an operation which reads the memory of the place and converts it to a value. This
+/// is a fundamentally *typed* operation. The nature of the value produced depends on the type of
+/// the conversion. Furthermore, there may be other effects: if the type has a validity constraint
+/// the place to value conversion might be UB if the validity constraint is not met.
+///
+/// **Needs clarification:** Ralf proposes that place to value conversions not have side-effects.
+/// This is what is implemented in miri today. Are these the semantics we want for MIR? Is this
+/// something we can even decide without knowing more about Rust's memory model?
+///
+/// A place to value conversion on a place that has its variant index set is not well-formed.
```

Comment:
Why is this needed? Miri at least is fine with such places being loaded from. The result is basically the same as loading without the variant index set (but the variant index is 'carried over' to the type information of the created value).

Reply:
I vaguely remember seeing some code that would ICE if this was the case, but now can't find it. In any case, I can't think of a good reason to do this, and being able to check this in the validator seems like it could catch some bugs. I can change this to a "needs clarification" if this is controversial though.

Reply:
It seems a bit too arbitrary to be canonical, so yeah let's move it to "needs clarification".